File: Linked_List/817.Linked_List_Components/C++/817.Linked_List_Components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode * head, vector<int> & nums) 
    {
        //====================================//
        // Input type:                        //
        //   - ListNode * (head list-node)    //
        //   - vector<int> ref. (nums vector) //
        // Return type:                       //
        //   - int (result summary counter)   //
        //====================================//

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record dictionary, hashmap /////
        unordered_map<int, bool> record_map;

        ///// Result summary counter /////
        int res_cnter = 0;


        /*Hashmap-based loop traversal*/
        ///// Step 1: Record indexed-value informations with dictionary, hashmap /////
        for (int nums_idx = 0; (nums_idx < len_nums); (nums_idx++)) //Whole
        {
            record_map[nums[nums_idx]] = true; //Keep updating, recording

        } //Whole

        ///// Step 2: Looped-traversal with recorded dictionary, hashmap /////
        while (head != nullptr) //Whole
        {
            ///// Check if the current list-node value existed or not /////
            if (! record_map[(head->val)]) { head = (head->next); continue; }

            (res_cnter++); //Keep updating, accumulating

            while (((head->next) != nullptr) && record_map[((head->next)->val)]) //Part
            {
                head = (head->next); //Keep updating, traversing

            } //Part

            head = (head->next); //Keep updating, traversing

        } //Whole

        return res_cnter;
    }
};

File: Linked_List/817.Linked_List_Components/Python/817.Linked_List_Components.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def numComponents(self, head, nums):
        """
        :type head: Optional[ListNode]
        :type nums: List[int]
        :rtype: int
        """
        #================================#
        # Hashmap-based traversal method #
        #================================#

        ############
        #Initialize
        ##### Length of nums array #####
        len_nums = len(nums)

        ##### Record dictionary/hashmap #####
        record_dict = {}

        ##### Result summary counter #####
        res_cnter = 0


        ##############################
        #Hashmap-based loop traversal
        ##### Step 1: Record indexed-value informations with dictionary/hashmap #####
        for nums_idx in range(len_nums):
            record_dict[nums[nums_idx]] = True #Keep updating/recording

        ##### Step 2: Looped-traversal with recorded dictionary/hashmap #####
        while (head):

            ##### Check if the current list-node value existed or not #####
            if (head.val not in record_dict):
                head = head.next #Keep updating/traversing

                continue

            res_cnter += 1 #Keep updating/accumulating

            while (head.next and ((head.next).val in record_dict)):
                head = head.next #Keep updating/traversing

            head = head.next #Keep updating/traversing

        return res_cnter
